test/e2e/storage/csi_mock_volume.go

```diff
@@ -0,0 +1,462 @@
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+type cleanupFuncs func()
+
+const (
+	csiNodeLimitUpdateTimeout  = 5 * time.Minute
+	csiPodUnschedulableTimeout = 2 * time.Minute
+)
+
+var _ = utils.SIGDescribe("CSI mock volume", func() {
+	type mockDriverSetup struct {
+		cs           clientset.Interface
+		config       *testsuites.PerTestConfig
+		testCleanups []cleanupFuncs
+		pods         []*v1.Pod
+		pvcs         []*v1.PersistentVolumeClaim
+		sc           map[string]*storage.StorageClass
+		driver       testsuites.TestDriver
+		nodeLabel    map[string]string
+		provisioner  string
+	}
+	var (
+		m               mockDriverSetup
+		attachable      bool
+		deployCRD       bool
+		podInfoVersion  *string
+		scName          string
+		attachLimit     int
+		nodeSelectorKey string
+	)
+
+	f := framework.NewDefaultFramework("csi-mock-volumes")
+
+	init := func() {
+		m = mockDriverSetup{
+			cs: f.ClientSet,
+			sc: make(map[string]*storage.StorageClass),
+		}
+		csics := f.CSIClientSet
+		var err error
+
+		m.driver = drivers.InitMockCSIDriver(deployCRD, attachable, podInfoVersion, attachLimit)
+		config, testCleanup := m.driver.PrepareTest(f)
+		m.testCleanups = append(m.testCleanups, testCleanup)
+		m.config = config
+		m.provisioner = config.GetUniqueDriverName()
+
+		if nodeSelectorKey != "" {
+			framework.AddOrUpdateLabelOnNode(m.cs, m.config.ClientNodeName, nodeSelectorKey, f.Namespace.Name)
+			m.nodeLabel = map[string]string{
+				nodeSelectorKey: f.Namespace.Name,
+			}
+		}
+
+		if deployCRD {
+			err = waitForCSIDriver(csics, m.config.GetUniqueDriverName())
+			framework.ExpectNoError(err, "Failed to get CSIDriver : %v", err)
+			m.testCleanups = append(m.testCleanups, func() {
+				destroyCSIDriver(csics, m.config.GetUniqueDriverName())
+			})
+		}
+	}
+
+	createPod := func() (*storage.StorageClass, *v1.PersistentVolumeClaim, *v1.Pod) {
+		By("Creating pod")
+		var sc *storagev1.StorageClass
+		if dDriver, ok := m.driver.(testsuites.DynamicPVTestDriver); ok {
+			sc = dDriver.GetDynamicProvisionStorageClass(m.config, "")
+		}
+		nodeName := m.config.ClientNodeName
+		scTest := testsuites.StorageClassTest{
+			Name:         m.driver.GetDriverInfo().Name,
+			Provisioner:  sc.Provisioner,
+			Parameters:   sc.Parameters,
+			ClaimSize:    "1Gi",
+			ExpectedSize: "1Gi",
+		}
+		if scName != "" {
+			scTest.StorageClassName = scName
+		}
+		nodeSelection := testsuites.NodeSelection{
+			// The mock driver only works when everything runs on a single node.
+			Name: nodeName,
+		}
+		if len(m.nodeLabel) > 0 {
+			nodeSelection = testsuites.NodeSelection{
+				Selector: m.nodeLabel,
+			}
+		}
+		class, claim, pod := startPausePod(f.ClientSet, scTest, nodeSelection, f.Namespace.Name)
+		if class != nil {
+			m.sc[class.Name] = class
+		}
+		if claim != nil {
+			m.pvcs = append(m.pvcs, claim)
+		}
+		if pod != nil {
+			m.pods = append(m.pods, pod)
+		}
+		return class, claim, pod
+	}
+
+	resetSharedVariables := func() {
```

Comment:
sorry this is really hard to read and figure out when/where variables are set.  Can these variables be recreated every iteration instead of needing to reset them after every test?

Reply:
I moved them into a struct and it gets explicitly created at each iteration.